core/src/main/java/com/linecorp/armeria/client/RedirectingClient.java

```diff
@@ -573,24 +567,87 @@ CompletableFuture<HttpResponse> responseFuture() {
+                                                                                  request.headers().path(),
+                                                                                  request.method());
+                redirectSignatures.add(originalSignature);
+                isAddedOriginalSignature = true;
+            }
+
+            final RedirectSignature signature = new RedirectSignature(nextReqTarget.scheme(),
+                                                                      nextReqTarget.authority(),
+                                                                      nextReqTarget.pathAndQuery(),
+                                                                      nextMethod);
+            return redirectSignatures.add(signature);
+        }
+
         String originalUri() {
             if (originalUri == null) {
                 originalUri = buildUri(ctx, request.headers());
             }
             return originalUri;
         }
 
-        boolean addRedirectUri(HttpMethod method, String redirectUri) {
-            if (redirectUris == null) {
-                redirectUris = LinkedListMultimap.create();
+        Set<RedirectSignature> redirectSignatures() {
+            // Always called after addRedirectSignature is called.
+            assert redirectSignatures != null;
+            return redirectSignatures;
+        }
+
+        Set<String> redirectUris() {
+            // Always called after addRedirectSignature is called.
+            assert redirectSignatures != null;
+            return redirectSignatures.stream()
+                                     .map(RedirectSignature::uri)
+                                     .collect(ImmutableSet.toImmutableSet());
+        }
+    }
+
+    @VisibleForTesting
+    static class RedirectSignature {
+        private final String protocol;
+        private final String authority;
+        private final String pathAndQuery;
+        private final HttpMethod method;
+
+        RedirectSignature(String protocol, String authority, String pathAndQuery, HttpMethod method) {
+            this.protocol = protocol;
+            this.authority = authority;
+            this.pathAndQuery = pathAndQuery;
+            this.method = method;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(protocol, authority, pathAndQuery, method);
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (obj == this) {
+                return true;
+            }
+            if (!(obj instanceof RedirectSignature)) {
+                return false;
             }
-            return redirectUris.put(method, redirectUri);
+
+            final RedirectSignature that = (RedirectSignature) obj;
+            return protocol.equals(that.protocol) &&
+                   authority.equals(that.authority) &&
+                   pathAndQuery.equals(that.pathAndQuery) &&
+                   method == that.method;
```

Comment:
How about ordering the comparison in more cost-effective way? For example, we could compare `pathAndQuery` and `authority` before others because they are more likely to change, although it may not always be the case.